addons/hr_expense/tests/test_expenses.py

```diff
@@ -1098,6 +1098,19 @@ def test_corner_case_expense_reported_cannot_be_zero(self):
         # CASE 8: ALLOWS Setting the amounts to 0 while unlinking the expense sheet
         expense.write({'total_amount_currency': 0.0, 'total_amount': 0.0, 'sheet_id': False})
 
+        # CASE 9: FORBIDS Creating an expense with amounts to 0 from the list view in the expense sheet
+        expense_sheet = self.create_expense_report(values={'expense_line_ids': []})
+        expense_form = Form(self.env['hr.expense'].with_context({
+            'default_company_id': expense_sheet.company_id,
+            'default_employee_id': expense_sheet.employee_id,
+            'default_payment_mode': expense_sheet.payment_mode or 'own_account',
+            'check_total_amount_not_zero': True,
+            }))
+        expense_form.name = 'Test Name'
+        expense_form.total_amount_currency = 0.0
+        with self.assertRaises(UserError):
+            expense_sheet.expense_line_ids = [expense_form.save()]
```

Comment:
```suggestion
        with Form(expense_sheet) as sheet_form:
            with sheet_form.expense_line_ids.new() as expense_form:
                .....
                with self.assertRaises(UserError):
                      expense_form.save()   # An error is thrown when an expense with a total of zero is created , 
                .....
                expense_form.total_amount_currency = 10.0  # Able to save, as it's non-zero
```
The flow is using a form on hr.expense.sheet, not hr.expense (and JSYK, Form is a context manager so you can use with Form as, which has the additional benefits of saving when exiting)
See the tests for the split wizard for examples of how create/edit/remove lines inside a form

So I'd seen something more like this, it's a bit more clear IMO

Reply:
EDIT: stupid idea, a tour is better to check the context